File: research/compression/nengo_experiments/spikecoder/source.py
"""The momentum-rover Markov source.

Lifted byte-compatible from ``research/compression/validate.py`` so every spiking
result is asserted against the *same* numbers the paper and the numpy validators
use. A rover moves on the alphabet {U, D, L, R} with long-run move frequencies
``pi = (1/2, 1/4, 1/8, 1/8)`` but with *inertia*: with stickiness ``s`` it repeats
its last move, otherwise it draws a fresh move from ``pi``. Its transition matrix
is ``P_ij = s * 1[i=j] + (1-s) * pi_j`` -- a convex blend of the identity (pure
momentum) and the rank-one ``1 pi^T`` (pure memorylessness).

At s = 0.7: marginal entropy H(pi) = 1.7500 bits, entropy rate H(P) = 0.9782 bits,
mutual information I = 0.7718 bits (44% of the marginal cost the cycle can recover).
"""

import numpy as np

from .config import SEED

LABELS = ["U", "D", "L", "R"]
PI = np.array([1 / 2, 1 / 4, 1 / 8, 1 / 8])
S = 0.7


def momentum_chain(pi, s):
    """P_ij = s*delta_ij + (1-s)*pi_j  (the 'stay-or-resample' chain)."""
    pi = np.asarray(pi, float)
    n = len(pi)
    return s * np.eye(n) + (1.0 - s) * np.tile(pi, (n, 1))


def sample_stream(P, pi, n, seed=0):
    """Sample an n-symbol stream from the chain P started at pi."""
    rng = np.random.default_rng(seed)
    N = len(pi)
    x = np.empty(n, dtype=int)
    x[0] = rng.choice(N, p=pi)
    for t in range(1, n):
        x[t] = rng.choice(N, p=P[x[t - 1]])
    return x


class RoverSource:
    """A momentum-rover source with streaming helpers for Nengo input Nodes.

    Parameters
    ----------
    s : float
        Stickiness in [0, 1). Default 0.7 (the paper's worked operating point).
    pi : array
        Stationary move frequencies. Default (1/2, 1/4, 1/8, 1/8).
    seed : int
        Stream sampling seed. Default config.SEED (= 7, the validators' seed).
    """

    def __init__(self, s=S, pi=PI, seed=SEED):
        self.pi = np.asarray(pi, float)
        self.N = len(self.pi)
        self.seed = seed
        self.set_s(s)

    def set_s(self, s):
        """(Re)build the transition matrix for stickiness s (for non-stationary runs)."""
        self.s = float(s)
        self.P = momentum_chain(self.pi, self.s)
        return self

    # convenience alias used by the non-stationary capstone
    switch_s = set_s

    def sample(self, n, seed=None):
        """Sample an n-symbol stream (ints in 0..N-1)."""
        return sample_stream(self.P, self.pi, n, seed=self.seed if seed is None else seed)

    # ---- Nengo input-function builders -------------------------------------
    def context_input(self, symbols, window_dur):
        """Return f(t) -> one-hot of the *previous* symbol (the context c_t).

        Each symbol occupies a window of length ``window_dur`` seconds. During
        window t the context is the one-hot encoding of symbol x_{t-1} (for t=0,
        the all-zeros / uniform context).
        """
        symbols = np.asarray(symbols, int)
        N = self.N

        def f(t):
            k = int(t // window_dur)
            out = np.zeros(N)
            if 0 < k < len(symbols):
                out[symbols[k - 1]] = 1.0
            return out

        return f

    def outcome_input(self, symbols, window_dur):
        """Return f(t) -> one-hot of the *current* symbol x_t (the teacher y)."""
        symbols = np.asarray(symbols, int)
        N = self.N

        def f(t):
            k = int(t // window_dur)
            out = np.zeros(N)
            if 0 <= k < len(symbols):
                out[symbols[k]] = 1.0
            return out

        return f

File: crates/cogspike-core/src/coder/source.rs
//! The momentum-rover Markov source, byte-faithful to `spikecoder/source.py`.
//!
//! `P_ij = s*1[i=j] + (1-s)*pi_j` -- a convex blend of the identity (pure
//! momentum) and `1 pi^T` (pure memorylessness). See `research/compression/validate.py`.

use rand::SeedableRng as _;
use rand::distributions::{Distribution as _, WeightedIndex};
use rand::rngs::StdRng;

use super::constants::{PI, S, SEED};

/// Transition matrix `P_ij = s*delta_ij + (1-s)*pi_j` (the stay-or-resample chain).
#[must_use]
pub fn momentum_chain(pi: &[f64], s: f64) -> Vec<Vec<f64>> {
    let n = pi.len();
    (0..n)
        .map(|i| {
            pi.iter()
                .enumerate()
                .map(|(j, &pj)| (if i == j { s } else { 0.0 }) + (1.0 - s) * pj)
                .collect()
        })
        .collect()
}

/// Sample an `n`-symbol stream from chain `p` started at `pi`.
///
/// NOTE: this uses Rust's `StdRng` and does NOT reproduce numpy's PCG64 stream
/// byte-for-byte; only the statistics converge (empirical marginal -> `pi`). Exact
/// parity is asserted only on the closed-form quantities (the golden fixtures).
#[must_use]
pub fn sample_stream(p: &[Vec<f64>], pi: &[f64], n: usize, seed: u64) -> Vec<usize> {
    let mut out = Vec::with_capacity(n);
    if n == 0 {
        return out;
    }
    let mut rng = StdRng::seed_from_u64(seed);
    let pi_dist = WeightedIndex::new(pi).expect("pi must be a valid probability distribution");
    let row_dists: Vec<WeightedIndex<f64>> = p
        .iter()
        .map(|row| WeightedIndex::new(row).expect("each P row must be a valid distribution"))
        .collect();
    let mut prev = pi_dist.sample(&mut rng);
    out.push(prev);
    for _ in 1..n {
        let next = row_dists
            .get(prev)
            .expect("row index is a previously sampled symbol, in range")
            .sample(&mut rng);
        out.push(next);
        prev = next;
    }
    out
}

/// One-hot encoding of `symbol` in an `n`-dimensional vector (all-zeros if out of range).
#[must_use]
pub fn one_hot(symbol: usize, n: usize) -> Vec<f64> {
    let mut v = vec![0.0; n];
    if let Some(slot) = v.get_mut(symbol) {
        *slot = 1.0;
    }
    v
}

/// The momentum-rover source: stationary `pi`, stickiness `s`, transition `p`, seed.
#[derive(Clone, Debug)]
pub struct RoverSource {
    /// Stationary move frequencies.
    pub pi: Vec<f64>,
    /// Stickiness in `[0, 1)`.
    pub s: f64,
    /// Transition matrix `momentum_chain(pi, s)`.
    pub p: Vec<Vec<f64>>,
    /// Stream sampling seed.
    pub seed: u64,
}

impl RoverSource {
    /// Build a source with the given stickiness, stationary distribution, and seed.
    #[must_use]
    pub fn new(s: f64, pi: Vec<f64>, seed: u64) -> Self {
        let p = momentum_chain(&pi, s);
        Self { pi, s, p, seed }
    }

    /// The paper's worked rover: `pi = (1/2,1/4,1/8,1/8)`, `s = 0.7`, `seed = 7`.
    #[must_use]
    pub fn paper() -> Self {
        Self::new(S, PI.to_vec(), SEED)
    }

    /// (Re)build the transition matrix for a new stickiness `s` (non-stationary runs).
    pub fn set_s(&mut self, s: f64) {
        self.s = s;
        self.p = momentum_chain(&self.pi, s);
    }

    /// Alphabet size.
    #[must_use]
    pub fn n(&self) -> usize {
        self.pi.len()
    }

    /// Sample an `n`-symbol stream (see [`sample_stream`] for the RNG caveat).
    #[must_use]
    pub fn sample(&self, n: usize) -> Vec<usize> {
        sample_stream(&self.p, &self.pi, n, self.seed)
    }
}
